src/Storages/StorageFile.cpp

```diff
@@ -509,6 +498,31 @@ Pipe StorageFile::read(
             else
                 return metadata_snapshot->getColumns();
         };
+        // if we do multiple reads from pipe, we want to check if pipe is a regular file or a pipe
+        if (this_ptr->table_fd_was_used)
+        {
+            struct stat fd_stat;
+            if (fstat(this_ptr->table_fd, &fd_stat) == -1)
+            {
+                throw Exception("Cannot stat table file descriptor, inside " + this_ptr->getName(), ErrorCodes::CANNOT_STAT);
+            }
+            if (S_ISREG(fd_stat.st_mode))
+            {
+                if (this_ptr->table_fd_init_offset < 0)
+                {
+                    throw Exception("File descriptor isn't seekable, inside " + this_ptr->getName(), ErrorCodes::CANNOT_SEEK_THROUGH_FILE);
+                }
+                if (lseek(this_ptr->table_fd, this_ptr->table_fd_init_offset, SEEK_SET) < 0)
+                {
+                    throw Exception("File descriptor isn't seekable, inside " + this_ptr->getName(), ErrorCodes::CANNOT_SEEK_THROUGH_FILE);
+                }
+            }
+            else
+            {
+                // else if fd is not a regular file, then throw an Exception
+                throw Exception("Cannot read from a pipeline twice", ErrorCodes::CANNOT_READ_FROM_FILE_DESCRIPTOR);
```

Comment:
BTW why do you call this pipeline not pipe?

Reply:
Changed error message